packages/poaps/src/PoapsClient.ts

```diff
@@ -90,4 +103,143 @@ export class PoapsClient {
+  }
+
+  /**
+   * Retrieves the claim code information for a specific QR hash.
+   * @param {string} qr_hash - The QR hash.
+   * @returns {Promise<GetClaimCodeResponse>} The claim code response.
+   */
+  async getClaimCode(qr_hash: string): Promise<GetClaimCodeResponse> {
+    return await this.tokensApiProvider.getClaimCode(qr_hash);
+  }
+
+  /**
+   * Gets the status of a claim using its unique ID.
+   * @param {string} queue_uid - The unique ID of the claim.
+   * @returns {Promise<MintingStatus>} The status of the claim.
+   */
+  async getClaimStatus(queue_uid: string): Promise<MintingStatus> {
+    const claimStatusResponse = await this.tokensApiProvider.claimStatus(
+      queue_uid,
+    );
+    return claimStatusResponse.status;
+  }
+
+  /**
+   * Waits for a claim's status to move out of the 'IN_PROCESS' or 'PENDING' states.
+   * @param {string} queue_uid - The unique ID of the claim.
+   * @returns {Promise<MintingStatus>} The final status of the claim.
+   */
+  async waitClaimStatus(queue_uid: string): Promise<MintingStatus> {
+    let status: MintingStatus = MintingStatus.IN_PROCESS;
+    while (
+      status === MintingStatus.IN_PROCESS ||
+      status === MintingStatus.PENDING
+    ) {
+      try {
+        status = await this.getClaimStatus(queue_uid);
+        await sleep(1000);
+      } catch {
+        // do nothing
+      }
+    }
+    return status;
+  }
+
+  /**
+   * Initiates an asynchronous claim process and returns a unique queue ID.
+   * @param {WalletClaimtInput} input - The claim input details.
+   * @returns {Promise<string>} The unique queue ID.
+   */
+  async claimAsync(input: WalletClaimtInput): Promise<string> {
+    const secret = await this.getCodeSecret(input.qr_hash);
+
+    const response = await this.tokensApiProvider.postClaimCode({
+      address: input.address,
+      qr_hash: input.qr_hash,
+      secret: secret,
+      sendEmail: false,
+    });
+
+    return response.queue_uid;
+  }
+
+  /**
+   * Initiates a synchronous claim process. It waits for the claim to process and
+   * then returns the associated POAP.
+   * @param {WalletClaimtInput} input - The claim input details.
+   * @returns {Promise<POAP>} The associated POAP once the claim is complete.
+   * @throws {FinishedWithError} If the claim finishes with an error.
+   */
+  async claimSync(input: WalletClaimtInput): Promise<POAP> {
+    const queue_uid = await this.claimAsync(input);
+
+    const status = await this.waitClaimStatus(queue_uid);
+
+    if (status === MintingStatus.FINISH) {
+      let getCodeResponse = await this.getClaimCode(input.qr_hash);
+
+      while (getCodeResponse.result == null) {
+        await sleep(1000);
+        getCodeResponse = await this.getClaimCode(input.qr_hash);
+      }
+      return (
+        await this.fetch({
+          limit: 1,
+          offset: 0,
+          ids: [getCodeResponse.result.token],
+        })
+      ).items[0];
+    }
+
+    throw new FinishedWithError(input.qr_hash);
```

Comment:
Why does it fails here? What would the value of `status` be? Add a reason, maybe log something.

In other subject, code is better understood if the last statement is the return happy path and all errors are thrown inside conditionals. /cc @alavarello because was his fight, ;)